base/applications/rapps/installed.cpp

```diff
@@ -12,149 +12,226 @@
+        lpKeyName,
+        NULL,
+        NULL,
+        NULL,
+        &dwSize) != ERROR_SUCCESS)
+    {
+        return FALSE;
+    }
 
-    dwType = REG_SZ;
-    dwSize = MAX_PATH * sizeof(WCHAR);
-    if (RegQueryValueExW(ItemInfo->hSubKey,
-                         bModify ? szModify : szUninstall,
-                         NULL,
-                         &dwType,
-                         (LPBYTE) szPath,
-                         &dwSize) != ERROR_SUCCESS)
+    // allocate buffer.
+    // attention: dwSize is size in bytes, and RegQueryValueExW does not guarantee the terminating null character.
+    String.GetBuffer(dwSize + sizeof(WCHAR));
+
+    // query the value
+    if (RegQueryValueExW(hSubKey,
+        lpKeyName,
+        NULL,
+        NULL,
+        (LPBYTE)String.GetBuffer(),
+        &dwSize) != ERROR_SUCCESS)
     {
+        String.ReleaseBuffer();
+        String.Empty();
         return FALSE;
     }
+    String.GetBuffer()[dwSize / sizeof(WCHAR)] = L'0'; // ensure zero terminated
+    String.ReleaseBuffer();
+    return TRUE;
+}
 
-    return StartProcess(szPath, TRUE);
+BOOL CInstalledApplicationInfo::UninstallApplication(BOOL bModify)
+{
+    return StartProcess(bModify ? szModifyPath : szUninstallString, TRUE);
 }
 
-BOOL EnumInstalledApplications(INT EnumType, BOOL IsUserKey, APPENUMPROC lpEnumProc, PVOID param)
+LSTATUS CInstalledApplicationInfo::RemoveFromRegistry()
 {
-    DWORD dwSize = MAX_PATH, dwType, dwValue;
-    BOOL bIsSystemComponent, bIsUpdate;
-    ATL::CStringW szParentKeyName;
-    ATL::CStringW szDisplayName;
-    INSTALLED_INFO Info;
-    HKEY hKey;
-    LONG ItemIndex = 0;
+    ATL::CStringW szFullName = L"Software\\Microsoft\\Windows\\CurrentVersion\\Uninstall\\" + szKeyName;
 
-    Info.hRootKey = IsUserKey ? HKEY_CURRENT_USER : HKEY_LOCAL_MACHINE;
+    return RegDeleteKeyW(IsUserKey ? HKEY_CURRENT_USER : HKEY_LOCAL_MACHINE, szFullName);
+}
 
-    if (RegOpenKeyW(Info.hRootKey,
-                    L"Software\\Microsoft\\Windows\\CurrentVersion\\Uninstall",
-                    &hKey) != ERROR_SUCCESS)
-    {
-        return FALSE;
-    }
+BOOL CInstalledApps::Enum(INT EnumType, APPENUMPROC lpEnumProc, PVOID param)
+{
+    FreeCachedEntries();
+
+    HKEY RootKeyEnum[2] = { HKEY_CURRENT_USER ,HKEY_LOCAL_MACHINE };
 
-    while (RegEnumKeyExW(hKey, ItemIndex, Info.szKeyName.GetBuffer(MAX_PATH), &dwSize, NULL, NULL, NULL, NULL) == ERROR_SUCCESS)
+    // loop 2 times for both HKEY_CURRENT_USER and HKEY_LOCAL_MACHINE
+    for (int i = 0; i < 2; i++)
     {
-        Info.szKeyName.ReleaseBuffer();
-        if (RegOpenKeyW(hKey, Info.szKeyName.GetString(), &Info.hSubKey) == ERROR_SUCCESS)
+        DWORD dwSize = MAX_PATH;
+        HKEY hKey, hSubKey;
+        LONG ItemIndex = 0;
+        ATL::CStringW szKeyName;
+        //Info.hRootKey = IsUserKey ? HKEY_CURRENT_USER : HKEY_LOCAL_MACHINE;
+
+        if (RegOpenKeyW(RootKeyEnum[i],
+            L"Software\\Microsoft\\Windows\\CurrentVersion\\Uninstall",
```

Comment:
When running as 32 bit app, this is redirected to `Computer\HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\Uninstall`,
you will have to read 2 extra keys (Which was done in `GetInstalledVersion` before, using the `KEY_WOW64_32KEY` and `KEY_WOW64_64KEY` flags)

Reply:
ok, I think I fixed this now~